crates/sparrow-transforms/src/transform_pipeline.rs

```diff
@@ -0,0 +1,264 @@
+                        kind: unsupported.into()
+                    })
+                }
+            };
+            transforms.push(transform);
+            input_step = step;
+        }
+        Ok(Self {
+            partitions: Partitioned::default(),
+            transforms,
+            sink,
+        })
+    }
+}
+
+impl Pipeline for TransformPipeline {
+    fn initialize(&mut self, tasks: Partitioned<TaskRef>) {
+        self.partitions = tasks
+            .into_iter()
+            .map(|task| TransformPartition {
+                is_closed: AtomicBool::new(false),
+                inputs: Mutex::new(VecDeque::new()),
+                task,
+            })
+            .collect();
+    }
+
+    fn add_input(
+        &self,
+        input_partition: Partition,
+        input: usize,
+        batch: Batch,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        error_stack::ensure!(
+            input == 0,
+            PipelineError::InvalidInput {
+                input,
+                input_len: 1
+            }
+        );
+        let partition = &self.partitions[input_partition];
+        error_stack::ensure!(
+            !partition.is_input_closed(),
+            PipelineError::InputClosed {
+                input,
+                input_partition
+            }
+        );
+
+        partition.add_input(batch);
+        scheduler.schedule(partition.task.clone());
+        Ok(())
+    }
+
+    fn close_input(
+        &self,
+        input_partition: Partition,
+        input: usize,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        error_stack::ensure!(
+            input == 0,
+            PipelineError::InvalidInput {
+                input,
+                input_len: 1
+            }
+        );
+        let partition = &self.partitions[input_partition];
+        error_stack::ensure!(
+            !partition.is_input_closed(),
+            PipelineError::InputClosed {
+                input,
+                input_partition
+            }
+        );
+
+        // Don't close the sink here. We may be currently executing a `do_work`
+        // loop, in which case we need to allow it to output to the sink before
+        // we close it.
+        partition.close_input();
+        scheduler.schedule(partition.task.clone());
+
+        Ok(())
+    }
+
+    fn do_work(
+        &self,
+        input_partition: Partition,
+        scheduler: &mut dyn Scheduler,
+    ) -> error_stack::Result<(), PipelineError> {
+        let partition = &self.partitions[input_partition];
+
+        let Some(batch) = partition.pop_input() else {
+            error_stack::ensure!(
+                partition.is_input_closed(),
+                PipelineError::illegal_state("scheduled without work")
+            );
+            return self.sink.close_input(input_partition, scheduler);
+        };
```

Comment:
Hm, so the logic here is that if a worker calls `do_work`, but the partitioned has no input scheduled, we'll close the sink (consumer) partition. However, if this pipeline's input is not already closed, then we'll error, because we expected this partition to already be closed by...the producing pipeline?

Reply:
Roughly. If we get here, then *this pipeline* asked to be scheduled.

1. It may have been requested in `add_input`, in which case there should be input.
2. It may have been requested in `close_input` to signal things are closed.
3. It may be the previous `do_work` realizing there is more to do.

In any of those cases, there should either be a batch to process or we should be all closed up, or there is a bug.

Ideally, (2) could detect "nothing is running, go ahead and close directly", but that is tricky. Since it just happens at the end, it doesn't seem worth guarding against.